src/v/storage/segment.cc

```diff
@@ -514,14 +516,20 @@ ss::future<append_result> segment::do_append(const model::record_batch& b) {
           b.header())));
     }
     const auto start_physical_offset = _appender->file_byte_offset();
+    const auto expected_end_physical = start_physical_offset
+                                       + b.header().size_bytes;
+
     _generation_id++;
+
+    // inflight index. trimmed on every dma_write in appender
+    _inflight.emplace(expected_end_physical, b.last_offset());
+
     // proxy serialization to segment_appender
     auto write_fut = _appender->append(b).then(
-      [this, &b, start_physical_offset] {
+      [this, &b, start_physical_offset, expected_end_physical] {
```

Comment:
> When appender advances stable offset, a callback for advancing stable
stable offset at segment level is also called,
`segment::advance_stable_offset`. This method uses the inflight writes
to determine how far the segment stable offset can be advanced.

Ahh, I see now. Great find. I'm very surprised this never happened in Debug builds where the scheduler randomizes the order of execution of runnable continuations, and that it took very fast storage to trigger.

> Because
the last inflight write wasn't registered yet, the stable offset can get
stuck.

This part is pretty cool, that normally we could drop the upcall, but it would only really matter if no new appends arrived promptly.

Reply:
>I'm very surprised this never happened in Debug builds

+1 also I'm wondering if this is somehow more likely with the new dirty tracking logic in the batch cache, though nothing really jumps out.

Reply:
Generally it was masked by code in `segment::do_flush`

```cpp
ss::future<> segment::do_flush() {
    _generation_id++;
    if (!_appender) {
        return ss::make_ready_future<>();
    }
    auto o = _tracker.dirty_offset;
    auto fsize = _appender->file_byte_offset();
    return _appender->flush().then([this, o, fsize] {
        // never move committed offset backward, there may be multiple
        // outstanding flushes once the one executed later in terms of offset
        // finishes we guarantee that all previous flushes finished.
        _tracker.committed_offset = std::max(o, _tracker.committed_offset);
        // ########## HERE ##########
        _tracker.stable_offset = _tracker.committed_offset;
        // ########## HERE ##########
        _reader->set_file_size(std::max(fsize, _reader->file_size()));
        clear_cached_disk_usage();
    });
}
```

However, very important!, this also does contain a bug and with concurrent flushes data can become missing https://github.com/redpanda-data/redpanda/pull/17039